### PR TITLE
Middleware, JOI, Errors handling

### PR DESCRIPTION
**Pull request by DianaBondarenko**
## Results
There were added middlewares for user's authentication, validation of headers and products in order, classes for some errors an error handler.

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -3,18 +3,28 @@ const productRouter = require('./routes/productRouter.js');
...
 const app = express();
 app.use(express.json());
 app.use(express.urlencoded({extended:false}));
 const PORT = 3000;
 
+app.use(morgan('dev'));
+
 app.use('/products', productRouter);
-app.use('/order', orderRouter);
+app.use('/order', validator.headers(authDTO), authMiddleWare, validator.body(orderDTO), orderRouter);
```
Comment (ghost) [inline app.js]: Better provide middlewares in the `orderRouter`. Also, please, specify request type inside `orderRouter` and `productRouter` files. Right now you're using `.use()` everywhere. I guess, there should be at least `.get()`.

File: models/orderModel.js [hunk 2]
```diff
@@ -23,23 +25,38 @@ class OrderModel {
...
-            await nodeMailer.sendMail(userInfo, orderInfo);
-        } catch (er) {
-            return this.getError([],er.message) ;
-            console.log('Some error in db:', er)
-        }
-        return this.getOk();
+            if (availability.success === 'false') return availability;
+            const {data} = await this.insertItems(products, user.id);
+            await nodeMailer.sendMail(user, data);
+        } catch (er) {return er}
```
Comment (ghost) [inline models/orderModel.js]: Should we handle error here?

**Review by ghost — COMMENTED** [review]
Please, add newline to the end of files.

File: app.js (hunk 1 above)
Comment (kolesnicknick) [reply app.js]: sure. perfect is post here, as it is POST request

File: app.js [hunk 3]
```diff
@@ -3,18 +3,28 @@ const productRouter = require('./routes/productRouter.js');
...
+const authMiddleWare = require('./common/middleware/authMiddleWare.js');
+const authDTO = require('./common/middleware/authDTO.js');
+const orderDTO = require('./common/middleware/orderDTO.js');
+const validator = require('express-joi-validation').createValidator({});
+const errorHandler = require('./common/middleware/errorHandler.js');
 
 const app = express();
 app.use(express.json());
 app.use(express.urlencoded({extended:false}));
 const PORT = 3000;
```
Comment (kolesnicknick) [inline app.js]: Use environment variable instead of just const

File: app.js [hunk 4]
```diff
@@ -3,18 +3,28 @@ const productRouter = require('./routes/productRouter.js');
...
 
+app.use(morgan('dev'));
+
 app.use('/products', productRouter);
-app.use('/order', orderRouter);
+app.use('/order', validator.headers(authDTO), authMiddleWare, validator.body(orderDTO), orderRouter);
+app.use(errorHandler);
 app.use((req, res)  => {
     res.status(404).send('Not Found');
```
Comment (kolesnicknick) [inline app.js]: We should raise error here

File: models/orderModel.js [hunk 5]
```diff
@@ -23,23 +25,38 @@ class OrderModel {
...
+    //     return this.getOk();
+    // }
     async addOrder(products, user) {
         try {
-            const res = await this.createUser(user);
-            if (res.status === 'error') return res;
-            const userInfo = res.data[0];
-            const valid = this.validateProducts(products)
-            if(valid.status === 'error') return valid;
+            if (user.isAuthenticated === 'false') return new Unauthorized('User is unauthorized');
```
Comment (kolesnicknick) [inline models/orderModel.js:48]: we should throw new Error here I guess. returning errors is not good

File: models/orderModel.js [hunk 6]
```diff
@@ -62,9 +79,9 @@ class OrderModel {
             if (rows.length <= 0) notFoundProducts.push({id: products[i].id})
             else if (rows[0].amount < products[i].count) notEnoughProducts.push(products[i]);
         }
-        if (notFoundProducts.length > 0) return this.getError(notFoundProducts, 'Products are not found');
-        if (notEnoughProducts.length > 0) return this.getError(notEnoughProducts, 'Not enough products');
-        return this.getOk();
+        if (notFoundProducts.length > 0) return new NotFound('Products are not found');
+        if (notEnoughProducts.length > 0) return new NotFound('Not enough products');
+        return this.getSuccess();
```
Comment (kolesnicknick) [inline models/orderModel.js]: same here, we should better throw Error than return it

File: models/orderModel.js [hunk 7]
```diff
@@ -123,9 +141,16 @@ class OrderModel {
...
+    //         data: data,
+    //         message: ''
+    //     }
+    // }
+    getSuccess(data = []) {
         return {
-            status: 'ok',
+            success: 'true',
             data: data,
             message: ''
```
Comment (kolesnicknick) [inline models/orderModel.js]: message still is not very useful here

**Pull request merged**